### ISSUE TITLE
CPack package inconsistencies

### ISSUE THREAD

**Issue by theopolis** [osquery/osquery#5769]
### For DEB

EDIT (9/7@10AM ET): Fixed the current paths. Before I was working on a dirty checkout.

- [x] ~~The CPack version may install into `/usr/etc` instead of `/etc`~~
- [x] `postinst` should not be included on the filesystem
- [x] We should be including the current `certs.pem` from OpenSSL (#5773)
- [x] The installed binary should be stripped
- [x] The package name should follow normal DEB convention
- [x] The revision should be "1.linux"
- [x] (4.0.2) Investigate building -dbg version (needs ELF build-id)
- [x] No license or homepage is set (nice-to-have)

### For RPM

- [x] The package name should follow normal RPM convention
- [x] License is still set to BSD
- [x] The revision should be "1.linux"
- [x] (4.0.2) Investigate producing -debuginfo version  (needs ELF build-id)

### For PKG

- [X] ~Investigate removing `Resources` (welcome, readme, license)~
- [x] Revert to legacy `com.facebook.osquery`
- [x] (4.0.2) Investigate producing -debug version

### For TGZ

- [x] The `TGZ` generator adds a prefix directory (would need a workaround)

### For WIX

TODO

### All

- [x] Investigate adding changelog to the packages
- [x] Restore COPYING to augeas lenses directory

**Comment by Smjert** [comment]
I'm a bit confused about the wording here and the relative changes in the PR.
With "Previously" and "Currently" I guess you mean 3.3.2 vs now? 
If so, I've followed that package format and if you look at the pre-release packages we have released, they do install in the same path as 3.3.2.

Normally you don't force any prefix on the user and let him choose through CPACK_PACKAGING_INSTALL_PREFIX which already has a default to `/usr`.
Also, when you specify absolute path in the `install()` directive, those are not relocatable, which means that they'll be installed in that path.
There's only one way to change that path and it's using DESTDIR while installing (which afaik is what CPack does to install in his private folder to then package).

For the TGZ, you have to enable the [CPACK_SET_DESTIR](https://cmake.org/cmake/help/latest/variable/CPACK_SET_DESTDIR.html) option.

**Comment by theopolis** [comment]
> I'm a bit confused about the wording here and the relative changes in the PR.
With "Previously" and "Currently" I guess you mean 3.3.2 vs now?
If so, I've followed that package format and if you look at the pre-release packages we have released, they do install in the same path as 3.3.2.
Have you had a different experience?

Sorry, I made a mistake when I compared "previous"==3.3.x and "current"=master. I removed the absolute `/` for a few of the `install()` `DESTINATION` paths to resolve CMake warnings about relocatability. 

> Normally you don't force any prefix on the user and let him choose through CPACK_PACKAGING_INSTALL_PREFIX which already has a default to /usr.
Also, when you specify absolute path in the install() directive, those are not relocatable, which means that they'll be installed in that path.

I am not familiar with CPack best practices. Should we use absolutes like `DESTINATION /etc/default`? This is what is in master now, and I was attempting to change it in #5770 to be relative. I will follow your guidance here.

> For the TGZ, you have to enable the CPACK_SET_DESTIR option.

Thanks for the tip, I will take a look.

Let me follow up in another comment about the `certs.pem`.

**Comment by theopolis** [comment]
I updated the top-post to reflect reality and to be more clear about "previously" and "currently". Thanks for pointing out that bug.

**Comment by theopolis** [comment]
@Smjert, the reason we include a `certs.pem` is because some operating systems do not include one by default, for example macOS's OpenSSL was not supported on several versions and thus the certificate bundle was outdated.

I think the goal is to provide a sane default, but that is not represented in the code. Ideally the code should look for well-known locations to find a bundle then fall-back to our provided PEM. We should follow up on this in another issue and PR. Until then I'd like to not break compatibility and continue to provide an updated bundle with the package.

**Comment by theopolis** [comment]
On MacOS / productbuild we do not want to use a component build. CPack behaves oddly but the resulting "package" can be found at:
```
./_CPack_Packages/Darwin/productbuild/osquery-$VERSION/Contents/Packages/osquery
```

**Comment by Smjert** [comment]
> On MacOS / productbuild we do not want to use a component build. CPack behaves oddly but the resulting "package" can be found at:
> 
> ```
> ./_CPack_Packages/Darwin/productbuild/osquery-$VERSION/Contents/Packages/osquery
> ```

What you mean it behaves oddly?

**Comment by theopolis** [comment]
I'm reading the code for the non-component build: https://github.com/Kitware/CMake/blob/master/Source/CPack/cmCPackProductBuildGenerator.cxx#L53

This creates a pkg with the following BOM:
```
macos10:build vagrant$ lsbom $(pkgutil --bom /build/build/_CPack_Packages/Darwin/productbuild/osquery-4.0.0-94-g9777cef5/Contents/Packages/osquery) | grep -v -e packs -e lenses
.	40755	0/0
./Contents	40755	0/0
./Contents/Packages	40755	0/0
./Contents/Packages/osquery	100644	0/0	0	4294967295
./Contents/Packages/scripts	40755	0/0
./private	40755	0/0
./private/var	40755	0/0
./private/var/log	40755	0/0
./private/var/log/osquery	40755	0/0
./private/var/osquery	40755	0/0
./private/var/osquery/certs	40755	0/0
./private/var/osquery/certs/certs.pem	100644	0/0	224085	3329711587
./private/var/osquery/com.osquery.osqueryd.conf	100644	0/0	186	2265089030
./private/var/osquery/com.osquery.osqueryd.plist	100644	0/0	563	1904581733
./private/var/osquery/osquery.example.conf	100644	0/0	6215	2264020213
./usr	40755	0/0
./usr/local	40755	0/0
./usr/local/bin	40755	0/0
./usr/local/bin/osqueryctl	100755	0/0	3600	693803279
./usr/local/bin/osqueryd	100755	0/0	13136536	1050878180
./usr/local/bin/osqueryi	120755	0/0	8	2588742340	osqueryd
```

And I do not see how to remove the `Contents` section.

**Comment by theopolis** [comment]
For `apt` I am noticing an issue / inconsistency.

The `deb` package generated by CPack:
```
dpkg-deb -I osquery_4.0.1_1.linux.amd64.deb
 new Debian package, version 2.0.
 size 7111636 bytes: control archive=6876 bytes.
      43 bytes,     2 lines      conffiles            
     286 bytes,    11 lines      control              
   16858 bytes,   234 lines      md5sums              
     431 bytes,    18 lines   *  postinst             #!/bin/sh
 Architecture: amd64
 Depends: libc6 (>=2.12), zlib1g
 Description: osquery is an operating system instrumentation toolchain.
 Maintainer: osquery@osquery.io
 License: Apache-2.0 OR GPL-2.0-only
 Package: osquery
 Priority: extra
 Section: default
 Version: 4.0.1-1.linux
 Installed-Size: 54123
```
The attributes are provided in alphabetical order.

When I use `freight` to create the `apt` repo this builds a `Packages` with the following
```
[...]
Package: osquery
Version: 3.3.2-1.linux
License: BSD
Vendor: Facebook
Architecture: amd64
Maintainer: osquery@osquery.io
Installed-Size: 37420
Depends: libc6 (>=2.12), zlib1g
Section: default
Priority: extra
Homepage: https://osquery.io
Description: osquery is an operating system instrumentation toolchain.
Filename: pool/deb/main/o/osquery/osquery_3.3.2-1.linux_amd64.deb
MD5sum: 1cbf2872e87ad3c5b3a50caedd556fd1
SHA1: 3b99daea6ea706472e14a7a5a0d97d966a76a33e
SHA256: 6c54a55df1feaf410064bc87696f9829d0787fb5b9c7beabeb4f5cca7ce6c3fb
SHA512: a2daa2faccfd2c1d14f7fa14844da58a81173db74e0ed1dca5ab908ad368a57201330b5001766be106f3ea8c7eee83a503afe2ae350daa36a1268c3295ed8232
Size: 10359520

Architecture: amd64
Depends: libc6 (>=2.12), zlib1g
Description: osquery is an operating system instrumentation toolchain.
Maintainer: osquery@osquery.io
License: Apache-2.0 OR GPL-2.0-only
Package: osquery
Priority: extra
Section: default
Version: 4.0.1-1.linux
Installed-Size: 54123
Filename: pool/deb/main/o/osquery/osquery_4.0.1-1.linux_amd64.deb
MD5sum: b6db75a76f0886f96416652ea28ebe4f
SHA1: b2621742c18fdf304fdd42e0f145ab2c753a9800
SHA256: 7f3e58e5191d8ba579d0298a7cf625602da8881a1eef45dbb691d0d3ac1c92a1
SHA512: 8b076870537e7a615e5f87640b56ef2b72ed48c86df1e278902c2475d3207b4e973a44aa5b59a90ed15149cb98365816e39f9c2e3920b0e6174d496309214ba1
Size: 7112240
```

Example: https://osquery-packages.s3.amazonaws.com/deb/dists/deb/main/binary-amd64/Packages

And then `apt` does not pick up that `4.0.1-1.linux` is available.

**Comment by Smjert** [comment]
Just out of curiosity, why do we want to remove the Resources directory?
Wouldn't it be good to support also the graphical install?

For the changelog file I think that's a matter of installing it through "install()".
If needed the RPM generator also have `CPACK_RPM_CHANGELOG_FILE` which says
```
May be used to embed a changelog in the spec file. The referred file will be read and directly put after the %changelog section.
```

Same for COPYING.

**Comment by theopolis** [comment]
> Just out of curiosity, why do we want to remove the Resources directory?
Wouldn't it be good to support also the graphical install?
It's mostly my opinion, but the user-experience of the graphical install with the non-component package is superior. It is simple and requires very few clicks. This is the UX of installing the published 4.0.2 package.

The only priority issue with macOS is setting the identifier correctly to `com.facebook.osquery`. We could change the package identifier but it will have unknown effects for a lot of deployments, I strongly suggest we keep the legacy ID.

**Comment by theopolis** [comment]
Complete 💃
